crates/ide/src/ide/assists/inline.rs

```diff
@@ -0,0 +1,89 @@
+use super::{AssistKind, AssistsCtx};
+use crate::def::{AstPtr, ResolveResult};
+use crate::TextEdit;
+use smol_str::ToSmolStr;
+use syntax::ast::AstNode;
+use syntax::{ast, match_ast};
+
+pub(super) fn inline(ctx: &mut AssistsCtx<'_>) -> Option<()> {
+    let file_id = ctx.frange.file_id;
+    let parse = ctx.db.parse(file_id);
+
+    let covering_node = ctx.covering_node::<ast::Ref>()?;
+    let covering_token = covering_node.token()?;
+    let ptr = AstPtr::new(covering_node.syntax());
+
+    let name_res = ctx.db.name_resolution(file_id);
+    let source_map = ctx.db.source_map(file_id);
+    let expr_id = source_map.expr_for_node(ptr)?;
+
+    let replacement = match name_res.get(expr_id)? {
+        &ResolveResult::Definition(name) => source_map.nodes_for_name(name).find_map(|ptr| {
+            let definition_node = ptr.to_node(&parse.syntax_node());
+            let definition_value = definition_node.ancestors().find_map(|n| {
+                match_ast! {
+                    match n {
+                        ast::AttrpathValue(path_value) => path_value.value(),
+                        _ => None,
+                    }
+                }
+            })?;
+
+            Some(definition_value)
+        }),
+        // Only let-ins are rewritable
+        _ => None,
+    }?;
+    let replacement_text = {
+        let node = AstNode::syntax(&replacement);
+        let do_parenthesize = matches!(&replacement, ast::Expr::Lambda(_));
```

Comment:
We actually do have a function to decide whether an expression can be safely inlined in another one without parenthesis: https://github.com/oxalica/nil/blob/b043bfe1f3f4c4be4b688e24c5ae96e81f525805/crates/syntax/src/ast.rs#L268

It is used like this:
https://github.com/oxalica/nil/blob/b043bfe1f3f4c4be4b688e24c5ae96e81f525805/crates/ssr/src/lib.rs#L97-L100